api/main.go

```diff
@@ -152,3 +153,195 @@ func (d *Dispatcher) Write(ctx context.Context, vl *ValueList) error {
+}
+
+// Untyped returns the specific value of ConfigValue without specifying its type, useful for functions like fmt.Printf
+// which can use variables with unknown types.
+func (v ConfigValue) Untyped() interface{} {
+	switch v.typ {
+	case configTypeString:
+		return v.s
+	case configTypeNumber:
+		return v.f
+	case configTypeBoolean:
+		return v.b
+	}
+	return nil
+}
+
+// Config represents one configuration block, which may contain other configuration blocks.
+type Config struct {
+	Key      string
+	Values   []ConfigValue
+	Children []Config
+}
+
+// Merge appends other's children to c's children.
+// Returns an error if Key or any Values differ.
+func (c *Config) Merge(other *Config) error {
+	panic("Not yet implemented")
+}
+
+// Unmarshal applies the configuration from a Config to an arbitrary struct.
+func (c *Config) Unmarshal(v interface{}) error {
+	// Sanity check value of the interface
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return fmt.Errorf("can only unmarshal to a non-nil pointer") // TODO: better error message or nil if preferred
+	}
+
+	// If the target supports unmarshalling let it
+	if u, ok := v.(Unmarshaler); ok {
+		return u.UnmarshalConfig(v)
+	}
+
+	drv := rv.Elem() // get dereferenced value
+	drvk := drv.Kind()
+
+	// If config has child configs we can only unmarshal to a struct or slice of structs
+	if len(c.Children) > 0 {
+		if drvk != reflect.Struct && (drvk != reflect.Slice || drv.Elem().Kind() != reflect.Struct) {
+			return fmt.Errorf("cannot unmarshal a config with children except to a struct or slice of structs")
+		}
+	}
+
+	switch drvk {
+	case reflect.Invalid, reflect.Array, reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.UnsafePointer:
+		return fmt.Errorf("cannot unmarshal into type %s", drv.Type())
+	case reflect.Struct:
+		// Unmarshal values from config
+		if err := storeStructConfigValues(c.Values, drv); err != nil {
+			return fmt.Errorf("while unmarshalling values into %s: %s", drv.Type(), err)
+		}
+		for i := range c.Children {
+			// If a config has children but the struct has no corresponding field, or the corresponding field is an
+			// unexported struct field the child is ignored without notice.
+			if field := drv.FieldByName(c.Children[i].Key); field.IsValid() && field.CanInterface() {
+				fieldPtr := field.Addr().Interface()
+
+				if err := c.Children[i].Unmarshal(fieldPtr); err != nil {
+					return fmt.Errorf("while unmarshalling child config with key %s: %s", c.Children[i].Key, err)
+				}
+
+			}
+		}
+		return nil
+	case reflect.Slice:
+		switch rv.Elem().Kind() {
+		case reflect.Struct:
+			// Create a temporary Value of the same type as dereferenced value, then get a Value of the same type as
+			// its elements. Unmarshal into that Value and append the temporary Value to the original.
+			tv := reflect.New(drv.Type()).Elem()
+			if err := storeStructConfigValues(c.Values, tv); err != nil {
```

Comment:
I was expecting a recursive call here, e.g.:

```go
if err := c.Unmarshal(tv.Addr().Interface()); err != nil {
  return fmt.Errorf("unmarshaling into the newly allocated element failed …")
}
// append to drv
```

Reply:
That makes sense, will implement it that way.